libcnb-cargo/src/main.rs

```diff
@@ -223,3 +379,332 @@ fn calculate_dir_size(path: impl AsRef<Path>) -> io::Result<u64> {
+                                    buildpack_project.id, local_dependency_id));
+                }
+            };
+
+            if !buildpack_projects_seen.contains(&local_dependency.id) {
+                buildpack_projects_seen.insert(&local_dependency.id);
+                buildpack_projects_to_compile.push(local_dependency.clone());
+            }
+        }
+        if !buildpack_projects_seen.contains(&buildpack_project.id) {
+            buildpack_projects_seen.insert(&buildpack_project.id);
+            buildpack_projects_to_compile.push(buildpack_project.clone());
+        }
+    }
+
+    buildpack_projects_to_compile
+}
+
+fn get_cargo_metadata(buildpack_dir: &Path) -> Metadata {
+    match MetadataCommand::new()
+        .manifest_path(&buildpack_dir.join("Cargo.toml"))
+        .exec()
+    {
+        Ok(cargo_metadata) => cargo_metadata,
+        Err(error) => {
+            fail_with_error(format!("Could not obtain metadata from Cargo: {error}"));
+        }
+    }
+}
+
+fn get_current_dir() -> PathBuf {
+    match env::current_dir() {
+        Ok(current_dir) => current_dir,
+        Err(io_error) => {
+            fail_with_error(format!("Could not determine current directory: {io_error}"));
+        }
+    }
+}
+
+fn get_workspace_root() -> PathBuf {
+    let cargo = env::var("CARGO")
+        .map(PathBuf::from)
+        .ok()
+        .unwrap_or_else(|| PathBuf::from("cargo"));
+
+    let mut locate_project = Command::new(cargo);
+    locate_project.args(["locate-project", "--workspace", "--message-format", "plain"]);
+
+    match locate_project.output() {
+        Ok(output) => {
+            let output = String::from_utf8_lossy(&output.stdout);
+            parent_dir(&PathBuf::from(output.trim()))
+        }
+        Err(error) => {
+            fail_with_error(format!("Could not locate project root: {error}"));
+        }
+    }
+}
+
+fn find_buildpack_directories(buildpack_workspace: &BuildpackWorkspace) -> Vec<PathBuf> {
+    let paths = match glob(
+        &buildpack_workspace
+            .root
+            .join("**/buildpack.toml")
+            .to_string_lossy(),
+    ) {
+        Ok(paths) => paths,
+        Err(error) => fail_with_error(format!(
+            "Failed to glob buildpack.toml files in {}: {error}",
+            buildpack_workspace.root.to_string_lossy()
+        )),
+    };
+
+    let exclude_target_pattern = match glob::Pattern::new(
+        &buildpack_workspace
+            .target_dir
+            .join("**/buildpack.toml")
+            .to_string_lossy(),
+    ) {
+        Ok(pattern) => pattern,
+        Err(error) => fail_with_error(format!(
+            "Invalid glob pattern for {}: {}",
+            buildpack_workspace.target_dir.to_string_lossy(),
+            error
+        )),
+    };
+
+    paths
+        .filter_map(Result::ok)
+        .filter(|path| !exclude_target_pattern.matches(&path.to_string_lossy()))
+        .map(|path| parent_dir(&path))
+        .collect()
```

Comment:
i guess you could do that.  my approach was to halt the process vs. ignoring this unlikely error but i don't really have a strong preference for either since this should never really happen.